### ISSUE TITLE
No DCD_EVENT_UNPLUGGED event for STM32 FSdev

### ISSUE THREAD

**Issue by mauriciocirelli** [hathach/tinyusb#1476]
### Operating System

Windows 10

### Board

Custom based on STM32H7B3

### Firmware

portable/st/stm32_fsdev/dcd_stm32_fsdev.c

### What happened ?

Dear,

I have followed the example for CDC/MSC using FreeRTOS and have got pretty much everything working. Expect for the disconnection event: unmount callback (tud_umount_cb) is not being called for this platform and driver. Mount callback (tud_mount_cb) happens as expected.

I am new to this library, so maybe I have messed up something, but as everything else is working, I guess I maybe facing a bug somewhere. Looking into the code, the old synopsys driver evokes the DCD_EVENT_UNPLUGGED event, but the new stm32_fsdev does not.

Any help would be much appreciated.

### How to reproduce ?

Just connect the board using the new stm32_fsdev and look for the mount and unmount callbacks.

**Comment by tfx2001** [comment]
It may depend on the VBUS sensing, please check your schematic.

**Comment by tfx2001** [comment]
It seems that STM32H7B3 is using synopsys dwc2 driver, not use stm32_fsdev driver.

**Comment by mauriciocirelli** [comment]
> It seems that STM32H7B3 is using synopsys dwc2 driver, not stm32_fsdev driver.

Dear, thank you for pointing this out.
I have confirmed that I am using the synopsys dwc2 driver.

However, debugging `dcd_int_handler` function I have confirmed that the following code is not being executed when I unplug the device from the computer:

`dcd_event_bus_signal(rhport, DCD_EVENT_UNPLUGGED, true);`

I have enabled VBUS sensing in CubeMX and it is directly connected from the USB connector to PA9.

**Comment by hathach** [comment]
@mauriciocirelli indeed, the disconnect detection is a bit confusing on dwc2. The specs is a bit fuzzy. dwc2 driver may miss some logic here and there since it is a generic driver which work on all configuration of dwc2. The old stm synopsys driver is very specific to stm configured controller.

I will try to revise and test this out when having time.

**Comment by mauriciocirelli** [comment]
@hathach thank you for your reply. 
Detecting the disconnection event is important for us, because we have a FATFs running on this device and we need to unmount/re-mount it when connected/disconnected from USB. Otherwise, if our firmware tries to access the FATFs while Windows is accessing it from USB, the file system gets corrupted.

**Comment by hathach** [comment]
unfortunately, I am currently busy with other works. will check this out when possible. If you manage to get it work, please consider to make an PR to help others.

**Comment by BennyEvans** [comment]
I'm experiencing the same issue after recently updating tinyUSB and moving to the dwc2 driver. I'll have a quick look through the new driver to see if it's anything obvious but might revert to the old driver to keep the device working for now.

**Comment by BennyEvans** [comment]
Removing the line below seems to resolve the issue for me. May someone else please confirm this also works for them. I'm running in HS mode on an STM32h742 (custom hardware). https://github.com/hathach/tinyusb/blob/fd5bb6e5db8e8e997d66775e689cc73f149e7fc1/src/portable/synopsys/dwc2/dcd_dwc2.c#L525

**Comment by BennyEvans** [comment]
Switched to head and it also seems to be working with the fix above. Note however, there is another bug if running in HS mode with an STM32H7. You will need to modify a section in tusb_mcu.h to add `#define TUP_RHPORT_HIGHSPEED  1` - in this section:
https://github.com/hathach/tinyusb/blob/fd5bb6e5db8e8e997d66775e689cc73f149e7fc1/src/common/tusb_mcu.h#L158-L163

I can create a PR to add `#define TUP_RHPORT_HIGHSPEED  1` if needed.

**Comment by hathach** [comment]
h7 does not have built-in HS PHY, which is the `TUP_RHPORT_HIGHSPEED` is about. For board with external PHY, you need to have define `CFG_TUD_MAX_SPEED = OPT_MODE_HIGH_SPEED`

**Comment by BennyEvans** [comment]
@hathach my mistake! I see now. I'll give that a try. Thanks!

**Comment by mauriciocirelli** [comment]
@BennyEvans I can confirm that removing that line from the driver triggers the unmount event.
I am testing on v0.13, and the respective line is 514 instead of 525.

However, I am getting several triggers of this unmount event when I unplug the USB cable. Is it expected?

Also, I would like to try v0.14 to get things up to date, but it seems that there are some configurations that changed and I am stuck on check_dwc2 function.

This is the setup I have for v0.13:

```
#define CFG_TUSB_MCU                  OPT_MCU_STM32H7
#define CFG_TUSB_OS                   OPT_OS_FREERTOS
#define BOARD_DEVICE_RHPORT_SPEED     OPT_MODE_FULL_SPEED
#define BOARD_DEVICE_RHPORT_NUM       1
#define CFG_TUSB_RHPORT1_MODE         (OPT_MODE_DEVICE | OPT_MODE_FULL_SPEED)
#define CFG_TUSB_DEBUG                0
```

Is there any doc explaining the new equivalent settings?

**Comment by BennyEvans** [comment]
@mauriciocirelli I don't have the board with me at the moment so can't help. But will try to get to it over the next few days.

**Comment by mauriciocirelli** [comment]
Dear @BennyEvans,

The issue is worse than at first... It is calling the _tud_umount_cb_ callback so many times that it is giving me a STACK OVERFLOW error at the TinyUSB task.

I had to work around it by setting a flag like this:

```
void tud_mount_cb(void)
{
  if(bIsMounted)
    return;

  // Do something here
  bIsMounted = true;
}

// Invoked when device is unmounted
void tud_umount_cb(void)
{
  if(!bIsMounted)
    return;

  bIsMounted = false;
  // Do something here
}
```

**Comment by benishor** [comment]
I am using a daisy seed board which runs a STM32H7 and I can confirm I too do not get the unplugged event. The hack described above does not work either.

**Comment by milabs** [comment]
There is no disconnection event for [STM32F407VGT](https://stm32-base.org/boards/STM32F407VGT6-STM32F4XX-M.html) with 2-wire (D+/D-) USB configuration.

**Comment by HiFiPhile** [comment]
> same for me (STM32H7) the fix doesn't work. :(
> 
> How can it be that it is not possible to (actively) check whether we are still connected?

Yeah it's frustrating, ~~guess we need to ask how did Synopsis made a such buggy controller~~

In some cases you can active the sof interrupt and monitor if it hasn't been fired for more than 1ms.

**Comment by hathach** [comment]
I think dwc2 ip list these cases more clearly in databook than stm document. We just haven't done it properly. I will try to look at this later when having time.

**Comment by dmarman** [comment]
I can confirm that `DCD_EVENT_UNPLUGGED` is not triggered for `STM32F0` (uses `FSDEV` instead of `DWC2`). I've tried some of the hacks above unsuccessfully.

Polling `tud_connected()` is not really reliable as sometimes when the cable is disconnected, still returns true.
STM32F0 doesn't have a VBUS, so I guess my only chance here is to change my hardware and manually read voltage from the USB connection.

**Comment by HiFiPhile** [comment]
## Update

Now we have a better understanding of DWC2 core.

`GOTGINT_SEDET` bit used by disconnection detection is set only when the **utmiotg_bvalid** signal is deasserted. To make DCD_EVENT_UNPLUGGED  event work <ins>**you must have VBus signal connected**</ins> (in which case B-peripheral session valid override is not used). Otherwise you can use `tud_suspend_cb()` instead.

In fact without VBus signal USB device has no way to know the difference between suspend and disconnection since voltage level is the same.

**Comment by Viatorus** [comment]
@HiFiPhile thank you for update. I’m not fully understood your explanation.

Could you please clarify whether this could be resolved:
- via software (e.g., TinyUSB or STM32 driver/configuration)
- via hardware (e.g., missing or incorrect bus signaling)
- or if it’s a limitation inherent to the STM32 USB peripheral

**Comment by brentmellor** [comment]
Thanks @HiFiPhile for the quick response on my issue ticket. [(3475)](https://github.com/hathach/tinyusb/issues/3475)

It appears the Nucleo board does have a VBUS sense line attached to PC2, but it is merely a GPIO and doesn't have an alternate function which connects it to the USB peripheral.
<img width="1924" height="1270" alt="Image" src="https://github.com/user-attachments/assets/464f6631-923d-4ef6-9cd3-28fc5d4968cb" />

Do you think that the DWC2 driver could be made to accept a pin interrupt to signal VBUS rise/fall and initiate the proper callbacks through TinyUSB? I'll be doing this externally for now, but it would be nice to have it integrated.

**Comment by brentmellor** [comment]
Or perhaps the driver could be modified to have conditions such as USB reset (when previously connected) also trigger the tu_unmount_cb()?
That would effectively be the same result, and function on devices without VBUS connected.

**Comment by HiFiPhile** [comment]
@Viatorus 

> Could you please clarify whether this could be resolved:
via software (e.g., TinyUSB or STM32 driver/configuration)
via hardware (e.g., missing or incorrect bus signaling)
or if it’s a limitation inherent to the STM32 USB peripheral

If you have Vbus connected then DCD_EVENT_UNPLUGGED event and tu_unmount_cb should work normally.

If you don't have Vbus then DCD_EVENT_UNPLUGGED and tu_unmount_cb won't work, you have to use tud_suspend_cb instead.

**Comment by brentmellor** [comment]
Thank you, @HiFiPhile.

I've connected the hardware VBUS input (PA9 on the STM32U575 parts) but I am not getting the tud_unmount_cb event. However I have paralleled the suspend callback to the same software triggers and it is working for our our case.

**Comment by HiFiPhile** [comment]
@brentmellor 

> I've connected the hardware VBUS input (PA9 on the STM32U575 parts) but I am not getting the tud_unmount_cb event.

Have you defined OTG_FS_VBUS_SENSE to disable session override ? Also I had to disable session override in dcd_dwc2.c

> Do you think that the DWC2 driver could be made to accept a pin interrupt to signal VBUS rise/fall and initiate the proper callbacks through TinyUSB? I'll be doing this externally for now, but it would be nice to have it integrated.

<details><summary>Details</summary>



It's possible by toggle B session override in GPIO interrupt, I did a successful experiment on U5A5-Nucleo (HS).

Note:
- STM32's DWC2 has a customized configuration register GCCFG, sometimes session override needs to be done in GCCFG instead of GOTGCTL
- On Nucleo board PC2's voltage after divider is too low, I had to connect PC2 to 5V_UCPD directly.
- Maybe something need to be changed on U575 (FS)

```diff
diff --git a/src/portable/synopsys/dwc2/dcd_dwc2.c b/src/portable/synopsys/dwc2/dcd_dwc2.c
index 44f7137f9..d605c2281 100644
--- a/src/portable/synopsys/dwc2/dcd_dwc2.c
+++ b/src/portable/synopsys/dwc2/dcd_dwc2.c
@@ -472,8 +472,8 @@ bool dcd_init(uint8_t rhport, const tusb_rhport_init_t* rh_init) {
   // Force device mode
   dwc2->gusbcfg = (dwc2->gusbcfg & ~GUSBCFG_FHMOD) | GUSBCFG_FDMOD;
 
-  // Clear A override, force B Valid
-  dwc2->gotgctl = (dwc2->gotgctl & ~GOTGCTL_AVALOEN) | GOTGCTL_BVALOEN | GOTGCTL_BVALOVAL;
+  // Clear A override
+  dwc2->gotgctl = (dwc2->gotgctl & ~GOTGCTL_AVALOEN) | GOTGCTL_BVALOEN;
 
 #if CFG_TUSB_MCU == OPT_MCU_STM32N6
   // No hardware detection of Vbus B-session is available on the STM32N6
@@ -780,7 +780,7 @@ static void handle_bus_reset(uint8_t rhport) {
     dwc2->epout[0].doeptsiz |= (3 << DOEPTSIZ_STUPCNT_Pos);
   }
 
-  dwc2->gintmsk |= GINTMSK_OEPINT | GINTMSK_IEPINT | GINTMSK_IISOIXFRM;
+  dwc2->gintmsk |= GINTMSK_OTGINT | GINTMSK_OEPINT | GINTMSK_IEPINT | GINTMSK_IISOIXFRM;
 }
 
 static void handle_enum_done(uint8_t rhport) {
@@ -1180,6 +1180,7 @@ void dcd_int_handler(uint8_t rhport) {
     const uint32_t otg_int = dwc2->gotgint;
 
     if (otg_int & GOTGINT_SEDET) {
+      dwc2->gintmsk &= ~GINTMSK_OTGINT;
       dcd_event_bus_signal(rhport, DCD_EVENT_UNPLUGGED, true);
     }
 

diff --git a/hw/bsp/stm32u5/boards/stm32u575nucleo/board.h b/hw/bsp/stm32u5/boards/stm32u575nucleo/board.h
index be037b68a..a1447e4da 100644
--- a/hw/bsp/stm32u5/boards/stm32u575nucleo/board.h
+++ b/hw/bsp/stm32u5/boards/stm32u575nucleo/board.h
@@ -55,6 +55,12 @@ extern "C"
 #define UART_TX_PIN GPIO_PIN_7
 #define UART_RX_PIN GPIO_PIN_8
 
+// VBUS GPIO
+#define VBUS_GPIO_PORT GPIOC
+#define VBUS_GPIO_PIN GPIO_PIN_2
+#define VBUS_GPIO_IRQn EXTI2_IRQn
+#define VBUS_GPIO_HANDLER EXTI2_IRQHandler
+
 //--------------------------------------------------------------------+
 // RCC Clock
 //--------------------------------------------------------------------+
diff --git a/hw/bsp/stm32u5/boards/stm32u5a5nucleo/board.h b/hw/bsp/stm32u5/boards/stm32u5a5nucleo/board.h
index 0785fb36b..81d0486fb 100644
--- a/hw/bsp/stm32u5/boards/stm32u5a5nucleo/board.h
+++ b/hw/bsp/stm32u5/boards/stm32u5a5nucleo/board.h
@@ -55,6 +55,12 @@ extern "C"
 #define UART_TX_PIN GPIO_PIN_9
 #define UART_RX_PIN GPIO_PIN_10
 
+// VBUS GPIO
+#define VBUS_GPIO_PORT GPIOC
+#define VBUS_GPIO_PIN GPIO_PIN_2
+#define VBUS_GPIO_IRQn EXTI2_IRQn
+#define VBUS_GPIO_HANDLER EXTI2_IRQHandler
+
 //--------------------------------------------------------------------+
 // RCC Clock
 //--------------------------------------------------------------------+
diff --git a/hw/bsp/stm32u5/family.c b/hw/bsp/stm32u5/family.c
index 26d72d6a0..9525dd1ac 100644
--- a/hw/bsp/stm32u5/family.c
+++ b/hw/bsp/stm32u5/family.c
@@ -89,6 +89,9 @@ void board_init(void) {
   __HAL_RCC_GPIOG_CLK_ENABLE();
   __HAL_RCC_GPIOH_CLK_ENABLE();
 
+  // Enable SYSCFG clock for EXTI
+  __HAL_RCC_SYSCFG_CLK_ENABLE();
+
   UART_CLK_EN();
 
   /* Enable Instruction cache */
@@ -156,6 +159,14 @@ void board_init(void) {
   GPIO_InitStruct.Alternate = GPIO_AF10_USB;
   HAL_GPIO_Init(GPIOA, &GPIO_InitStruct);
 
+  /* Configure VBUS GPIO Pin with EXTI */
+#ifdef VBUS_GPIO_PORT
+  GPIO_InitStruct.Pin = VBUS_GPIO_PIN;
+  GPIO_InitStruct.Mode = GPIO_MODE_IT_RISING_FALLING;  // Interrupt on both edges
+  GPIO_InitStruct.Pull = GPIO_NOPULL;
+  HAL_GPIO_Init(VBUS_GPIO_PORT, &GPIO_InitStruct);
+  HAL_NVIC_EnableIRQ(VBUS_GPIO_IRQn);
+#endif
 #ifdef USB_DRD_FS
   // STM32U535/STM32U545
 
@@ -188,7 +199,9 @@ void board_init(void) {
 
   // B-peripheral session valid override enable
   USB_OTG_FS->GOTGCTL |= USB_OTG_GOTGCTL_BVALOEN;
+  #ifndef VBUS_GPIO_PORT
   USB_OTG_FS->GOTGCTL |= USB_OTG_GOTGCTL_BVALOVAL;
+  #endif
   #endif // vbus sense
 
   /* Enable USB power on Pwrctrl CR2 register */
@@ -223,7 +236,9 @@ void board_init(void) {
 
   // B-peripheral session valid override enable
   USB_OTG_HS->GCCFG |= USB_OTG_GCCFG_VBVALEXTOEN;
+  #ifndef VBUS_GPIO_PORT
   USB_OTG_HS->GCCFG |= USB_OTG_GCCFG_VBVALOVAL;
+  #endif
 #endif // USB_OTG_FS
 }
 
@@ -286,3 +301,24 @@ void HardFault_Handler(void) {
 // -nostdlib/-nostartfiles.
 void _init(void) {
 }
+
+//--------------------------------------------------------------------+
+// EXTI Interrupt Handlers for VBUS Detection
+//--------------------------------------------------------------------+
+#ifdef VBUS_GPIO_PORT
+void VBUS_GPIO_HANDLER(void) {
+  __HAL_GPIO_EXTI_CLEAR_IT(VBUS_GPIO_PIN);
+  // VBUS rising edge
+  if (HAL_GPIO_ReadPin(VBUS_GPIO_PORT, VBUS_GPIO_PIN) == GPIO_PIN_SET) {
+#ifdef USB_OTG_FS
+    USB_OTG_FS->GOTGCTL |= USB_OTG_GOTGCTL_BVALOVAL;
+  } else {
+    USB_OTG_FS->GOTGCTL &= ~USB_OTG_GOTGCTL_BVALOVAL;
+#elif defined(USB_OTG_HS)
+    USB_OTG_HS->GCCFG |= USB_OTG_GCCFG_VBVALOVAL;
+  } else {
+    USB_OTG_HS->GCCFG &= ~USB_OTG_GCCFG_VBVALOVAL;
+#endif
+  }
+}
+#endif
```
</details>
